README.md
[![Build Status](https://scrutinizer-ci.com/g/zicht/itertools/badges/build.png?b=master)](https://scrutinizer-ci.com/g/zicht/itertools/build-status/master)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/zicht/itertools/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/zicht/itertools/?branch=master) 
[![Code Coverage](https://scrutinizer-ci.com/g/zicht/itertools/badges/coverage.png?b=master)](https://scrutinizer-ci.com/g/zicht/itertools/?branch=master)

# Zicht Iterator Tools Library
The Iterator Tools, or itertools for short, are a collection of
convenience tools to handle sequences of data such as arrays,
iterators, and strings.  Some of the naming and API is based on the
Python itertools.

[Examples](#examples)

Common operations include:
- [mapping](#mapping): `map` and `mapBy`
- [filtering](#filtering): `filter`, `difference`
- [sorting](#sorting): `sorted`
- [grouping](#grouping): `groupBy`
- [reducing](#reducing): `accumulate`, `collapse`, and `reduce`

## Usage
In order to use the available itertools filters/functions via Twig, simply add this service definition in your `config/services.yaml`

```yaml
Zicht\Itertools\twig\Extension:
  tags: ['twig.extension']
```

## Scripts
- unit test: `composer test`
- lint test: `composer lint`

## Example data
The examples below will use the following data to illustrate how
various Iterator tools work:

```php
$words = ['Useful', 'Goonies', 'oven', 'Bland', 'notorious'];
$numbers = [1, 3, 2, 5, 4];
$vehicles = [
    [
        'id' => 1,
        'type' => 'car', 
        'wheels' => 4, 
        'colors' => ['red', 'green', 'blue'], 
        'is_cool' => false, 
        'price' => 20000,
    ],
    [
        'id' => 2,
        'type' => 'bike', 
        'wheels' => 2, 
        'colors' => ['red', 'green', 'blue'], 
        'is_cool' => false, 
        'price' => 600,
    ],
    [
        'id' => 5,
        'type' => 'unicicle', 
        'wheels' => 1, 
        'colors' => ['red'], 
        'is_cool' => true, 
        'price' => 150,
    ],
    [
        'id' => 9,
        'type' => 'car', 
        'wheels' => 8, 
        'colors' => ['blue'], 
        'is_cool' => true, 
        'price' => 100000,
    ],
];
```

## Examples
With the example data above, this is how you could use itertools to get all unique colors of the cars in alphabetical order:

```php
use Zicht\Itertools\util\Filters;
use function Zicht\Itertools\iterable;

$vehicles = iterable($vehicles)
    ->filter(Filters::equals('car', 'type')) // {[vehicle...], [vehicle...]}
    ->map('colors') // {0: ['red', 'green', 'blue'], 1: ['blue']}
    ->collapse() // {0: 'red', 1: 'green', 2: 'blue', 3: 'blue'}
    ->unique() // {0: 'red', 1: 'green', 2: 'blue'}
    ->sorted(); // {2: 'blue', 1: 'green', 0: 'red'}
```

You can achieve the same in Twig:

```twig
{% for vehicle_color in vehicles
    |it.filter(it.filters.equals('car', 'type'))
    |it.map('colors')
    |it.collapse
    |it.unique
    |it.sorted
%}
    {{ vehicle_color }}
{% endfor %}
```

## Getter strategy
Many itertools can be passed a `$strategy` parameter.  This parameter
is used to obtain a value from the elements in the collection.  The
`$strategy` can be one of three things:

1. null, in which case the element itself is returned.  For example:

    ```php
    use function Zicht\Itertools\iterable;

    $result = iterable($words)->map(null);
    var_dump($result);
    // {0: 'Useful', 1: 'Goonies', 2: 'oven', 3: 'Bland', 4: 'notorious'}
    ```

    Or in Twig:

    ```twig
   {{ dump(word|it.map) }}
    ```

2. a closure, in which case the closure is called with the element
   value and key as parameters to be used to compute a return value.
   For example:

    ```php
    use function Zicht\Itertools\iterable;

    $getDouble = fn($value, $key) => 2 * $value;
    $result = iterable($numbers)->map($getDouble);
    var_dump($result);
    // {0: 2, 1: 6, 2: 4, 3: 10, 4: 8}
    ```

   Or in Twig:
    ```twig
   {{ dump(numbers|it.map(num => 2 * num)) }}
    ```

3. a string, in which case this string is used to create a closure
   that tries to find public properties, methods, or array indexes.
   For example:

    ```php
    use function Zicht\Itertools\iterable;

    $result = iterable($vehicles)->map('type');
    var_dump($result);
    // {0: 'car', 1: 'bike', 2: 'unicicle', 3: 'car'}
    ```

   Or in Twig:

    ```twig
   {{ dump(word|it.map) }}
    ```

   The string can consist of multiple dot separated words, allowing
   access to nested properties, methods, and array indexes.

   If one of the words in the string can not be resolved into an
   existing property, method, or array index, the value `null` will be
   returned.  For example:

    ```php
    use function Zicht\Itertools\iterable;

    $result = iterable($vehicles)->map('colors.2');
    var_dump($result);
    // {0: 'blue', 1: 'blue', 2: null, 3: null}
    ```

   Or in Twig:

    ```twig
   {{ dump(vehicles|it.map('colors.2')) }}
    ```

## Fluent interface
One way to use the Iterator Tools is to convert the array, Iterator,
string, etc into an `IterableIterator`.  This class provides a fluent
interface to all of the common operations.  For example:

```php
use function Zicht\Itertools\iterable;

$result = iterable($vehicles)->filter('is_cool')->mapBy('id')->map('type');
var_dump($result);
// {5: 'unicicle', 9: 'car'}
```

Or in Twig:

```twig
{{ dump(vehicles|it.filter('is_cool').mapBy('id').map('type')) }}
```

## Mapping
Mapping converts one collection into another collection of equal
length.  Using `map` allows manipulation of the elements while `mapBy`
allows manipulation of the collection keys.

For example, we can use a closure to create a title for each element
in `$vehicles`:

```php
use function Zicht\Itertools\iterable;

$getTitle = fn($value, $key) => sprintf('%s with %s wheels', $value['type'], $value['wheels']);
$titles = iterable($vehicles)->map($getTitle);
var_dump($titles);
// {0: 'car with 4 wheels', ..., 3: 'car with 8 wheels'}
```

Using the string [getter strategy](#getter-strategy) we can easily get
the types for each element in `$vehicles` mapped by the vehicle
identifiers.  For example:

```php
use function Zicht\Itertools\iterable;

$types = iterable($vehicles)->mapBy('id')->map('type');
var_dump($types);
// {1: 'car', 2: 'bike', 5: 'unicicle', 9: 'car'}
```

Or in Twig:

```twig
{{ dump(vehicles|it.mapBy('id').map('type')) }}
```


There are several common mapping closures available
in [mappings.php](src/Zicht/Itertools/util/Mappings.php).  Calling these
functions returns a closure that can be passed to `map` and `mapBy`.
For example:

```php
use Zicht\Itertools\util\Mappings;
use function Zicht\Itertools\iterable;

$lengths = iterable($words)->map(Mappings::length());
var_dump($lengths);
// {0: 6, 1: 3, 2: 4, 3: 5, 4: 9}
```

Or in Twig:

```twig
{{ dump(words|it.map(it.mappings.length)) }}
```

## Filtering
Filtering converts one collection into another, possibly shorter,
collection.  Using `filter` each element in the collection is
evaluated, the elements that are considered `empty` will be rejected,
while the elements that are not `empty` will be allowd to pass through
the filter.

For example, we can use a closure to determine if an element is
expensive, the `filter` will then only allow the expensive elements
through:

```php
use function Zicht\Itertools\iterable;

$isExpensive = fn($value, $key) => $value['price'] >= 10000;
$expensiveTypes = iterable($vehicles)->filter($isExpensive)->map('type');
var_dump($expensiveTypes);
// {1: 'car', 9: 'car'}
```

Or in Twig:

```twig
{{ dump(vehicles|it.filter(vehicle => vehicle.price >= 10000).map('type')) }}
```

Using the string [getter strategy](#getter-strategy) we can get only
the `$vehicles` that are considered to be cool.  For example:

```php
use function Zicht\Itertools\iterable;

$coolVehicleTypes = iterable($vehicles)->filter('is_cool')->map('type');
var_dump($coolVehicleTypes);
// {5: 'unicicle', 9: 'car'}
```

Or in Twig:

```twig
{{ dump(vehicles|it.filter('is_cool').map('type')) }}
```

There are several common filter closures available
in [filters.php](src/Zicht/Itertools/util/Filters.php).  Calling these
function returns a closure that can be passed to `filter`.  For
example:

```php
use Zicht\Itertools\util\Filters;
use function Zicht\Itertools\iterable;

$movieWords = iterable($words)->filter(Filters::in(['Shining', 'My little pony', 'Goonies']));
var_dump($movieWords);
// {1: 'Goonies'}
```

Or in Twig:

```twig
{{ dump(words|it.filter(it.filters.in(['Shining', "My little pony', 'Goonies'])) }}
```

## Sorting
`sorted` converts one collection into another collection of equal size
but with the elements possibly reordered.

For example, using the `null` [getter strategy](#getter-strategy),
which is the default, we will sort using the element values in
ascending order:

```php
use function Zicht\Itertools\iterable;

$ordered = iterable($numbers)->sorted();
var_dump($ordered);
// {0: 1, 2: 2, 1: 3, 4: 4, 3: 5}
```

Or in Twig:

```twig
{{ dump(numbers|it.sorted }}
```

The sorting algorithm will preserve the keys and is guaranteed to be
stable.  I.e. when elements are sorted using the same value, then the
sorted order is guaranteed to be the same as the order of the input
elements.  This is contrary to the standard PHP sorting functions.

Using the closure [getter strategy](#getter-strategy) the returned
value is used to determine the order.  The closure is called exactly
once per element, and the resulting values must be comparable.  For
example:

```php
use function Zicht\Itertools\iterable;

$getLower = fn($value, $key) => strtolower($value);
$ordered = iterable($words)->sorted($getLower);
var_dump($ordered);
// {3: 'Bland', 1: 'Goonies', 2: 'oven', 0: 'Useful', 4: 'notorious'};
```

The [mappings.php](src/Zicht/Itertools/util/Mappings.php) provides a
mapping closure which returns a random number.  This can be used to
sort a collection in a random order.  For example:

```php
use Zicht\Itertools\util\Mappings;
use function Zicht\Itertools\iterable;

$randomized = iterable($words)->sorted(Mappings::random());
var_dump($randomized);
// {... randomly ordere words ...}
```

Or in Twig:

```twig
{{ dump(words|it.sorted(it.mappings.random)) }}
```

## Grouping
`groupBy` converts one collection into one or more collections that
group the elements together on a specific criteria.

For example, using the string [getter strategy](#getter-strategy) we
can group all the `$vehicles` of the same type together:

```php
use function Zicht\Itertools\iterable;

$vehiclesByType = iterable($vehicles)->groupBy('type');
var_dump($vehiclesByType);
// {'bike': {1: [...]}, 'car': {0: [...], 3: [...]} 'unicicle': {2: [...]}}
```

Or in Twig:

```twig
{{ dump(vehicles|it.groupBy('type')) }}
```

Not that the original keys of the vehicles are still part of the
resulting groups, and the elements within each group keep the order
that they had in the input, i.e. it uses the stable sorting provided
by [`sorted`](#sorting).

## Reducing
`reduce` converts a collection into a single value by calling a
closure of two arguments cumulatively to the elements in the
collection, from left to right.

For example, without any arguments `reduce` will add all elements of
the collection together:

```php
use function Zicht\Itertools\iterable;

$sum = iterable($numbers)->reduce();
var_dump($sum);
// 15
```

Or in Twig:

```twig
{{ dump(numbers|it.reduce) }}
```

In the above example, the default closure that is used looks like this:

```php
public static function add($a, $b): \Closure
{
    return $a + $b;
}
```

Given that `$numbers` consists of the elements {1, 3, 2, 5, 4}, the
`add` closure is called four times:

```php
$sum = Reductions::add(Reductions::add(Reductions::add(Reductions::add((1, 3), 2), 5), 4));
var_dump($sum);
// 15
```

There are several common reduction closures available
in [reductions.php](src/Zicht/Itertools/util/Reductions.php).  Calling
these functions returns a closure that can be passed to `reduction`.
For example:

```php
use Zicht\Itertools\util\Reductions;
use function Zicht\Itertools\iterable;

$scentence = iterable($words)->reduce(Reductions::join(' - '));
var_dump($scentence);
// 'Useful - Goonies - oven - Bland - notorious'
```

Or in Twig:

```twig
{{ dump(words|it.reduce(it.reductions.join(' - ')) }}
```

Another common reduction is chaining multiple lists together into one list.
We call this process collapse.  This process can also be achieved using `reduce`
and `chain` together, however, because it is used frequently the `collapse` helper
makes its usage easier, for example:

```php
use function Zicht\Itertools\iterable;

$flat = iterable([['one', 'two'], ['three']])->collapse();
var_dump($flat);
// {0: 'one', 1: 'two', 0: 'three'}
```

Or in Twig:

```twig
{% set data = [['one', 'two'], ['three']] %}
{{ dump(data|it.collapse) }}
```

# Maintainer(s)
* Boudewijn Schoon <boudewijn@zicht.nl>
* Virginia Meijer <virginia@zicht.nl>


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]
### Added|Changed|Deprecated|Removed|Fixed|Security
Nothing so far

## 3.0.2 - 2022-08-04
### Added
- Prevent segfaults and exceptions due to debugging info

## 3.0.1 - 2021-11-18
### Changed
- Replaced PSR-0 autoloading in favor of PSR-4.
- Added upgrade guide for migrating itertools from 2.x to 3.0

## 3.0.0 - 2021-10-21
### Added
- Added support for php 8.
- Added support for Twig 3.
### Changed
- The argument to `iterable()` is no longer required, instead,
  an empty iterator is made when no arguments are given.
- The first argument of `->reduce()` is now a required closure.
### Removed
- Removed support for php 5.
- Removed support for Twig 1.
- Removed everything marked deprecated in the previous major release.
- Removed the option to enable or disable the legacy Twig filters and functions.

## 2.15.0 - 2021-04-08
### Added
- Support for changing the name in Twig
- Enable or disable the legacy Twig filters and functions

When the `zicht/framework-extra-bundle` is used to load the Twig extension,
the following allows you to configure these properties.

  ```yaml
  zicht_framework_extra:
    itertools:
        twig_name: it                   # use `it` (default)
        twig_enable_legacy_api: true    # enable legacy api (default)
  ```


## 2.14.0 - 2021-03-25
### Changed
- Marked many functions as deprecated, i.e. `filters\match` is now available as `Filters::match`.
  This is in preparation of the 3.0 release that will be compatible with php 8, which ships with a
  global `match` statement of its own.

  ```php
  // Deprecated
  // function from filters.php, mappings.php, and reductions.php
  use function Zicht\Iterable\reductions\join;
  iterable(['a', 'b', 'c'])->reduce(join(' - '));

  // New style
  use Zicht\Itertools\util\Reductions;
  iterable(['a', 'b', 'c'])->reduce(Reductions::join(' - '));
  ```

- Marked many twig filters and functions as deprecated, i.e. `[1, 2, 3]|filter` is now available as
  `[1, 2, 3]|it.filter` or `it([1, 2, 3]).filter`.
  This is because twig has been adding filters that use the same names as itertools does.

  ```twig
  {# Deprecated #}
  {{ entries|group_by('key') }}

  {# Everything was replaced by `it` #}
  {{ entries|it.groupBy('key') }}
  ```

## 2.13.0 - 2020-05-22
### Added
- Added `constant` mapping helper.

## 2.12.3 - 2020-02-14
### Fixed
- Small code improvement on how `before` and `after` convert ISO date strings.

## 2.12.2 - 2020-01-31
### Fixed
- The `before` and `after` filter helpers will now convert ISO date string when the expected
  value is a `DateTimeInterface`.

## 2.12.1 - 2020-01-08
### Fixed
- Linter fixes.

## 2.12.0 - 2020-01-08
### Added
- Added `cache` and `string` mapping helper.
- Added `before` and `after` filter helpers.

## 2.11.0 - 2019-04-29
### Added
- Added `CollapseIterator` usable from php or twig.

## 2.10.14 - 2019-04-23
### Fixed
- The `select` mapper now accepts either an array or an object as its first parameter,
  when an object is given then`select` will create an array with objects, when an array is
  given then `select` will create an array with arrays, as before.

## 2.10.13 - 2018-12-07
### Fixed
- Update `composer.lock`.
- Update code to conform with `zicht/standards-php` 3.4.0.

## 2.10.12 - 2018-11-12
### Fixed
- `ChainTrait`, `MapTrait`, and `ZipTrait` would indicate that they throw `\ReflectionException`,
  while this never happens.  This caused other code to give unnecessary warnings.

## 2.10.11 - 2018-11-07
### Fixed
- Changed the auto loader to only include the src directory.

## 2.10.10 - 2018-07-02
### Added
- Support for Twig v2.

## 2.10.9 - 2018-06-14
### Fixed
- The `values`, `toArray`, and `items` traits will now work recursively.
  This is useful when you have a nested iterator and also iterate over
  those iterators in a nested way (php does not properly support this).

  In the below example, the `.values` is used to created a *nested* array, i.e.
  both the `GroupByIterator` and the nested `GroupedIterator` are replaced by an
  array, this allows multiple iterations to occur at the same time.

  ```twig
    {% for grouped in items|group_by('EventId')|sorted('first.EventDateTime').values %}
       {% set first_item = grouped|first %}
       ...
    {% endfor %}
  ```

## 2.10.8 - 2017-12-14
### Added
- Added mappers `json_encode` and `json_decode`.

## 2.10.7 - 2017-11-09
### Fixed
- Reducing using `min` or `max` now supports `\DateTime` instances.

## 2.10.6 - 2017-08-31
### Fixed
- Added missing `$sort` parameter to the `group_by` twig filter extension.

## 2.10.5 - 2017-08-04
### Added
- Added a new filter helper `filters\match`.

## 2.10.4 - 2017-07-31
### Changed
- Calling `values()` on `GroupByIterator` will return the `values()` of every
  item in the group, instead of the `GroupedIterator` instances.

## 2.10.3 - 2017-06-30
### Added
- Added `$discardEmptyContainer` parameter to `mappings/select`.

## 2.10.2 - 2017-06-22
### Added
- Added `not` filter to `filters.php`
### Deprecated
- We should no longer use `not_in(...)` and instead use `not(in(...))`.

## 2.10.1 - 2017-06-21
### Added
- Added `$strategy` parameter to `mappings/select`.
- Added `$discardNull` parameter to `mappings/select`.

## 2.10.0 - 2017-06-01
### Added
- DifferenceTrait, this allows a call like $a->difference($b).
- IntersectionTrait, this allows a call like $a->intersection($b).
- first_key and FirstTrait::firstKey, allowing to get the key of the first element in the iterable.
- last_key and LastTrait::lastKey, allowing to get the key of the last element in the iterable.
### Changed
- Interface definitions for every trait, allowing for optimizations by not having to convert certain
  instances when it is known to be of the correct type already.

src/iter.php
<?php

namespace iter;

use ArrayIterator;
use Closure;
use ReflectionClass;
use InvalidArgumentException;
use Iterator;

/**
 * Transforms anything into an Iterator or throws an InvalidArgumentException
 *
 * > mixedToIterator([1, 2, 3])
 * 1 2 3
 *
 * > mixedToIterator('foo')
 * f o o
 *
 * @param array|string|Iterator $iterable
 * @return Iterator
 */
function mixedToIterator($iterable)
{
    // an array is *not* an instance of Traversable (as it is not an
    // object and hence can not 'implement Traversable')
    if (is_array($iterable)) {
        $iterable = new ArrayIterator($iterable);
    }

    // a string is considered iterable in Python
    if (is_string($iterable)) {
        $iterable = new StringIterator($iterable);
    }

    // by now it should be an Iterator, otherwise throw an exception
    if (!($iterable instanceof Iterator)) {
        throw new InvalidArgumentException('Argument $ITERABLE must be a Traversable');
    }

    return $iterable;
}

function mixedToClosure($closure)
{
    if (is_callable($closure)) {
        $closure = function () use($closure) {
            return call_user_func_array($closure, func_get_args());
        };
    }

    if (!($closure instanceof Closure)) {
        throw new InvalidArgumentException('Argument $KEYSTRATEGY must be a Closure');
    }

    return $closure;
}

/**
 * Try to transforms something into a keyStrategy Closure.
 *
 * @param string|Closure
 * @return Closure
 */
function mixedToKeyStrategy($keyStrategy)
{
    if (is_string($keyStrategy)) {
        $keyStrategy = function ($value) use ($keyStrategy) {
            if (is_object($value) && property_exists($value, $keyStrategy)) {
                return $value->{$keyStrategy};
            }

            if (is_callable(array($value, $keyStrategy))) {
                return call_user_func(array($value, $keyStrategy));
            }

            if (is_array($value) && array_key_exists($keyStrategy, $value)) {
                return $value[$keyStrategy];
            }

            return null;
        };
    }

    return mixedToClosure($keyStrategy);
}

/**
 * Make an iterator that returns accumulated sums.
 *
 * If the optional $func argument is supplied, it should be a string:
 * add, sub, mul, min, or max.  Or it can be a Closure taking two
 * arguments that will be used to instead of addition.
 *
 * > accumulate([1,2,3,4,5])
 * 1 3 6 10 15
 *
 * > accumulate(['one', 'two', 'three'], function ($a, $b) { return $a . $b; })
 * 'one' 'onetwo' 'onetwothree'
 *
 * @param array|string|Iterable $iterable
 * @param string|Closure $func
 * @return AccumulateIterator
 */
function accumulate($iterable, $func = 'add')
{
    if (is_string($func)) {
        switch ($func) {
        case 'add':
            $func = function ($a, $b) {
                if (!is_numeric($a)) {
                    throw new InvalidArgumentException('Argument $A must be numeric to perform addition');
                }
                if (!is_numeric($b)) {
                    throw new InvalidArgumentException('Argument $B must be numeric to perform addition');
                }
                return $a + $b;
            };
            break;
        case 'sub':
            $func = function ($a, $b) {
                if (!is_numeric($a)) {
                    throw new InvalidArgumentException('Argument $A must be numeric to perform subtraction');
                }
                if (!is_numeric($b)) {
                    throw new InvalidArgumentException('Argument $B must be numeric to perform subtraction');
                }
                return $a - $b;
            };
            break;
        case 'mul':
            $func = function ($a, $b) {
                if (!is_numeric($a)) {
                    throw new InvalidArgumentException('Argument $A must be numeric to perform multiplication');
                }
                if (!is_numeric($b)) {
                    throw new InvalidArgumentException('Argument $B must be numeric to perform multiplication');
                }
                return $a * $b;
            };
            break;
        case 'min':
            $func = function ($a, $b) {
                if (!is_numeric($a)) {
                    throw new InvalidArgumentException('Argument $A must be numeric to determine minimum');
                }
                if (!is_numeric($b)) {
                    throw new InvalidArgumentException('Argument $B must be numeric to determine minimum');
                }
                return $a < $b ? $a : $b;
            };
            break;
        case 'max':
            $func = function ($a, $b) {
                if (!is_numeric($a)) {
                    throw new InvalidArgumentException('Argument $A must be numeric to determine maximum');
                }
                if (!is_numeric($b)) {
                    throw new InvalidArgumentException('Argument $B must be numeric to determine maximum');
                }
                return $a < $b ? $b : $a;
            };
            break;
        }
    }

    if (!($func instanceof Closure)) {
        throw new InvalidArgumentException('Argument $FUNC must be a Closure or string \'add\'');
    }

    return new AccumulateIterator(mixedToIterator($iterable), $func);
}


/**
 * Make an iterator that returns elements from the first iterable
 * until it is exhausted, then prodeeds to the next iterable, until
 * all the iterables are exhausted.  Used for creating consecutive
 * sequences as a single sequence.
 *
 * Note that the keys of the returned ChainIterator follow 0, 1, etc,
 * regardless of the keys given in the iterables.
 *
 * > chain([1, 2, 3], [4, 5, 6])
 * 1 2 3 4 5 6
 *
 * > chain('ABC', 'DEF')
 * A B C D E F
 *
 * @param array|string|Iterator $iterable1
 * @param array|string|Iterator $iterable2
 * @param array|string|Iterator $iterableN
 * @return ChainIterator
 */
function chain(/* $iterable1, $iterable2, ... */)
{
    $iterables = array_map(function ($iterable) { return mixedToIterator($iterable); }, func_get_args());
    $reflectorClass = new ReflectionClass('iter\ChainIterator');
    return $reflectorClass->newInstanceArgs($iterables);
}

/**
 * Make an iterator that returns eventy spaced values starting with
 * number $start.
 *
 * > count(10)
 * 10 11 12 13 14 ...
 *
 * > count(2.5, 0.5)
 * 2.5 3.0 3.5 4.0 ...
 *
 * @param int|float $start
 * @param int|float $step
 * @return CountIterator
 */
function count($start = 0, $step = 1)
{
    if (!(is_int($start) or is_float($start))) {
        throw new InvalidArgumentException('Argument $START must be an integer or float');
    }

    if (!(is_int($step) or is_float($step))) {
        throw new InvalidArgumentException('Argument $STEP must be an integer or float');
    }

    return new CountIterator($start, $step);
}

/**
 * Make an iterator returning elements from the $iterable and saving a
 * copy of each.  When the iterable is exhausted, return elements from
 * the saved copy.  Repeats indefinately.
 *
 * > cycle('ABCD')
 * A B C D A B C D A B C D ...
 *
 * @param array|string|Iterator $iterable
 * @return CycleIterator
 */
function cycle($iterable)
{
    return new CycleIterator(mixedToIterator($iterable));
}

/**
 * Make an iterator returning values from $iterable and keys from
 * $keyStrategy.
 *
 * When $keyStrategy is a string, the key is obtained through one of
 * the following:
 * 1. $value->{$keyStrategy}, when $value is an object and
 *    $keyStrategy is an existing property,
 * 2. call $value->{$keyStrategy}(), when $value is an object and
 *    $keyStrategy is an existing method,
 * 3. $value[$keyStrategy], when $value is an array and $keyStrategy
 *    is an existing key,
 * 4. otherwise the key will default to null.
 *
 * Alternatively $keyStrategy can be a closure.  In this case the
 * $keyStrategy closure is called with each value in $iterable and the
 * key will be its return value.
 *
 * > $list = [['id'=>1, 'title'=>'one'], ['id'=>2, 'title'=>'two']]
 * > keyCallback('id', $list)
 * 1=>['id'=>1, 'title'=>'one'] 2=>['id'=>2, 'title'=>'two']
 *
 * @param string|Closure $keyStrategy
 * @param array|string|Iterator $iterable
 * @return KeyCallbackIterator
 */
function keyCallback($keyStrategy, $iterable)
{
    return new KeyCallbackIterator(mixedToKeyStrategy($keyStrategy), mixedToIterator($iterable));
}

/**
 * Make an iterator that applies $func to every value or $iterable.
 * If additional iterables are passed, $func must take that many
 * arguments and is applied to the values from all iterables in
 * parallel.
 *
 * With multiple iterables, the iterator stops when the shortest
 * iterable is exhausted.
 *
 * > $minimal = function ($value) { return min(3, $value); };
 * > map($minimal, [1, 2, 3, 4]);
 * 3 3 3 4
 *
 * > $average = function ($value1, $value2) { return ($value1 + $value2) / 2; };
 * > map($average, [1, 2, 3], [4, 5, 6]);
 * 2.5 3.5 4.5
 *
 * @param Closure $func
 * @param array|string|Iterator $iterable1
 * @param array|string|Iterator $iterable2
 * @param array|string|Iterator $iterableN
 * @return MapIterator
 */
function map($func /* $iterable1, $iterable2, ... */)
{
//    if (!($func instanceof Closure)) {
//        throw new InvalidArgumentException('Argument $FUNC must be a Closure');
//    }

    $iterables = array_map(function ($iterable) { return mixedToIterator($iterable); }, array_slice(func_get_args(), 1));
    $reflectorClass = new ReflectionClass('iter\MapIterator');
//    return $reflectorClass->newInstanceArgs(array_merge(array($func), $iterables));
    return $reflectorClass->newInstanceArgs(array_merge(array(mixedToKeyStrategy($func)), $iterables));
}

/**
 * Make an iterator that returns $mixed over and over again.  Runs
 * indefinately unless the $times argument is specified.
 *
 * > repeat(2)
 * 2 2 2 2 2 ...
 *
 * > repeat(10, 3)
 * 10 10 10
 *
 * @param mixed $mixed
 * @param null|int $times
 * @return RepeatIterator
 */
function repeat($mixed, $times = null)
{
    if (!(is_null($times) or (is_int($times) and $times >= 0))) {
        throw new InvalidArgumentException('Argument $TIMES must be null or a positive integer');
    }

    return new RepeatIterator($mixed, $times);
}
/**
 * Make an iterator that returns consecutive groups from the
 * $iterable.  Generally, the $iterable needs to already be sorted on
 * the same key function.
 *
 * When $keyStrategy is a string, the key is obtained through one of
 * the following:
 * 1. $value->{$keyStrategy}, when $value is an object and
 *    $keyStrategy is an existing property,
 * 2. call $value->{$keyStrategy}(), when $value is an object and
 *    $keyStrategy is an existing method,
 * 3. $value[$keyStrategy], when $value is an array and $keyStrategy
 *    is an existing key,
 * 4. otherwise the key will default to null.
 *
 * Alternatively $keyStrategy can be a closure.  In this case the
 * $keyStrategy closure is called with each value in $iterable and the
 * key will be its return value.
 *
 * The operation of groupby() is similar to the uniq filter in Unix.
 * It generates a break or new group every time the value of the key
 * function changes (which is why it is usually necessary to have
 * sorted the data using the same key function).  That behavior
 * differs from SQL's GROUP BY which aggregates common elements
 * regardless of their input order.
 *
 * > $list = [['type'=>'A', 'title'=>'one'], ['type'=>'A', 'title'=>'two'], ['type'=>'B', 'title'=>'three']]
 * > groupby('type', $list)
 * 'A'=>[['type'=>'A', 'title'=>'one'], ['type'=>'A', 'title'=>'two']] 'B'=>[['type'=>'B', 'title'=>'three']]
 *
 * @param string|Closure $keyStrategy
 * @param array|string|Iterator $iterable
 * @return GroupbyIterator
 */
function groupby($keyStrategy, $iterable)
{
    return new GroupbyIterator(mixedToKeyStrategy($keyStrategy), mixedToIterator($iterable));
}

/**
 * Make an iterator that returns the values from $iterable sorted by
 * $keyStrategy.
 *
 * When determining the order of two entries the $keyStrategy is used
 * twice, once for each value, and the results are used to determine
 * the order.
 *
 * When $reverse is true the order of the results are reversed.
 *
 * The sorted() function is guaranteed to be stable.  A sort is stable
 * if it guarantees not to change the relative order of elements that
 * compare equal.  this is helpful for sorting in multiple passes (for
 * example, sort by department, then by salary grade).  This also
 * holds up when $reverse is true.
 *
 * > $list = [['type'=>'B', 'title'=>'second'], ['type'=>'C', 'title'=>'third'], ['type'=>'A', 'title'=>'first']]
 * > sorted('type', $list)
 * ['type'=>'A', 'title'=>'first'] ['type'=>'B', 'title'=>'second']] ['type'=>'C', 'title'=>'third']
 *
 * @param string|Closure $keyStrategy
 * @param array|string|Iterator $iterable
 * @param boolean $reverse
 * @return SortIterator
 */
function sorted($keyStrategy, $iterable, $reverse = false)
{
    if (!is_bool($reverse)) {
        throw new InvalidArgumentException('Argument $REVERSE must be boolean');
    }
    return new SortIterator(mixedToKeyStrategy($keyStrategy), mixedToIterator($iterable), $reverse);
}

function filter($closure, $iterable)
{
    return new FilterIterator(mixedToClosure($closure), mixedToIterator($iterable));
}
